libr/bin/dwarf.c

```diff
@@ -190,6 +190,72 @@ static const char *dwarf_langs[] = {
 	[DW_LANG_Fortran08] = "Fortran08"
 };
 
+typedef struct {
+	const ut8 *buf;
+	st64 len;
+	bool big_endian;
+	size_t address_size;
+} data_xtor;
+
+void data_xtor_init(data_xtor *e, const ut8 *buf, st64 len, bool big_endian) {
+	e->buf = buf;
+	e->len = len;
+	e->big_endian = big_endian;
+}
+
+ut8 data_xtor_u8(data_xtor *e, st64 *off) {
+	ut8 ret = 0;
+	if (*off < e->len) {
+		ret = e->buf[*off];
+		*off += 1;
+	}
+	return ret;
+}
+
+ut16 data_xtor_u16(data_xtor *e, st64 *off) {
+	ut16 ret = 0;
+	if (*off + 2 <= e->len) {
+		ret = *(uut16*)(e->buf + *off);
+		if (R_SYS_ENDIAN != (int)e->big_endian)
+			ret = r_swap_ut16 (ret);
+		*off += 2;
+	}
+	return ret;
+}
+
+ut32 data_xtor_u32(data_xtor *e, st64 *off) {
+	ut32 ret = 0;
+	if (*off + 4 <= e->len) {
+		ret = *(uut32*)(e->buf + *off);
+		if (R_SYS_ENDIAN != (int)e->big_endian)
+			ret = r_swap_ut32 (ret);
+		*off += 4;
+	}
+	return ret;
+}
+
+ut64 data_xtor_u64(data_xtor *e, st64 *off) {
+	ut64 ret = 0;
+	if (*off + 8 <= e->len) {
+		ret = *(uut64*)(e->buf + *off);
+		if (R_SYS_ENDIAN != (int)e->big_endian)
+			ret = r_swap_ut64 (ret);
+		*off += 8;
+	}
+	return ret;
+}
+
+ut64 data_xtor_unsigned(data_xtor *e, st64 *off, ut8 byte_size) {
```

Comment:
this function can be simplified to use the r_read apis directly without all those helper functions